_posts/2017-08-23-auto-healing-foreman-proxy.md

```diff
@@ -0,0 +1,109 @@
+---
+layout: post
+title: Creating a auto healing Foreman Proxy
+author: Sean O'Keeffe
+tags:
+- foreman
+- katello
+- smart proxy
+- aws
+---
+
+Auto healing infrastructure allows scalability, predictability and automated recovery. This is the story of how we created an auto healing Foreman Smart Proxy with Packer, Terraform & Ansible on AWS.
+
+## Why?
+Why not? Defining all our infrastructure in code means we can recover from problems quickly and efficiently, not to mention the other benefits of that.
+- All of our other cloud applications and infrastructure are auto scaling and rebuilt every 21 days, from NTP & DNS servers to Splunk clusters and all the Applications & Servers they are there to support are built using this process. We don't want to differentiate from the standard.
+- No upgrades! We just re-build on the newer version.
+- No sessions! Disabling all user access, SSH & console, to our boxes allows simple deployment to cloud providers in our secure environment.
+
+
+## The process
+
+NOTE: We use Katello, so this guide uses Content Smart Proxies, a similar (probably simpler) process could also work for Foreman's Smart Proxies.
+
+### Base AMI
+
+We start with a Base AMI, this is a custom AMI that we produce based off the official AWS one, it includes custom configuration for hardening, NTP, DNS configuration & more, just basic stuff we want on every machine we deploy. You could replace this with one from the AWS Marketplace if you wanted to. Using Packer we took this base AMI to create a new "foreman-proxy" AMI. This just installs all the packages we need to run foreman-proxy, unregisters from the Katello server & SSH disabled, using the Shell Provisioner. None of our instances on AWS have SSH enabled or yum repos configured when running.
+
+```json
+<SNIP>
+
+"provisioners": [
+    {
+        "type": "shell",
+        "inline": [
+            "sed -i 's/console=ttyS0,115200n8 console=tty0/console=tty0 console=ttyS0,115200n8/' /etc/default/grub",
+            "sudo grub2-mkconfig -o /boot/grub2/grub.cfg",do subscription-manager clean",
+            "rpm -q katello-ca-consumer-katello.example.com 2>/dev/null || sudo rpm -Uvh http://katello.example.com/pub/katello-ca-consumer-latest.noarch.rpm",
```

Comment:
Won't that break if you ever decide to update the certs and your AMI has the older already installed?

Reply:
Yes, but we rebuild the AMI every 21 days anyway and could easier rebuild on-demand.

Reply:
I was more pointing at "why not use `yum install http://..` unconditionally"